BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/EnableBiometricLoginActivity.kt

```diff
@@ -0,0 +1,119 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+package com.example.biometricloginsample
+
+import android.content.Context
+import android.os.Bundle
+import android.util.Log
+import android.view.inputmethod.EditorInfo
+import androidx.activity.viewModels
+import androidx.appcompat.app.AppCompatActivity
+import androidx.biometric.BiometricManager
+import androidx.biometric.BiometricPrompt
+import androidx.core.widget.doAfterTextChanged
+import androidx.lifecycle.Observer
+import com.example.biometricloginsample.databinding.ActivityEnableBiometricLoginBinding
+
+class EnableBiometricLoginActivity : AppCompatActivity() {
+    private val TAG = "EnableBiometricLogin"
+    private lateinit var cryptographyManager: CryptographyManager
+    private val loginViewModel by viewModels<LoginWithPasswordViewModel>()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val binding = ActivityEnableBiometricLoginBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        binding.cancel.setOnClickListener { finish() }
+
+        loginViewModel.loginWithPasswordFormState.observe(this, Observer {
+            val loginState = it ?: return@Observer
+            when (loginState) {
+                is SuccessfulLoginFormState -> binding.authorize.isEnabled = loginState.isDataValid
+                is FailedLoginFormState -> {
+                    loginState.usernameError?.let { binding.username.error = getString(it) }
+                    loginState.passwordError?.let { binding.password.error = getString(it) }
+                }
+            }
+        })
+        binding.username.doAfterTextChanged {
+            loginViewModel.onLoginDataChanged(
+                binding.username.text.toString(),
+                binding.password.text.toString()
+            )
+        }
+        binding.password.doAfterTextChanged {
+            loginViewModel.onLoginDataChanged(
+                binding.username.text.toString(),
+                binding.password.text.toString()
+            )
+        }
+        binding.password.setOnEditorActionListener { _, actionId, _ ->
+            when (actionId) {
+                EditorInfo.IME_ACTION_DONE ->
+                    loginWithPassword(
+                        binding.username.text.toString(),
+                        binding.password.text.toString()
+                    )
+            }
+            false
+        }
+        binding.authorize.setOnClickListener {
+            loginWithPassword(binding.username.text.toString(), binding.password.text.toString())
+        }
+    }
+
+    private fun loginWithPassword(username: String, password: String) {
+        val succeeded = loginViewModel.login(username, password)
```

Comment:
This is good separation, but in a real app the login could take a long time, and so it shouldn't be called synchronously like this. Instead it should kick off `login` and it should use something like `LiveData` to report the user logged in, which the `Activity` can observe and update the UI based on that change.

Reply:
yup

Reply:
Please don't resolve the comment until the code change is completed. :)